lib/blockstore/file.js

```diff
@@ -0,0 +1,791 @@
+   * the hash is read from the block header.
+   * @private
+   * @param {Number} type - The type of block data
+   * @returns {Promise}
+   */
+
+  async _index(type) {
+    const {missing, filenos} = await this.check(type);
+
+    if (!missing)
+      return;
+
+    this.logger.info('Indexing block type %d...', type);
+
+    for (const fileno of filenos) {
+      const b = this.db.batch();
+      const filepath = this.filepath(type, fileno);
+      const data = await fs.readFile(filepath);
+      const reader = bio.read(data);
+      let magic = null;
+      let blocks = 0;
+
+      while (reader.left() >= 4) {
+        magic = reader.readU32();
+
+        // Move forward a byte from the last read
+        // if the magic doesn't match.
+        if (magic !== this.network.magic) {
+          reader.seek(-3);
+          continue;
+        }
+
+        let hash = null;
+        let position = 0;
+        let length = 0;
+
+        try {
+          length = reader.readU32();
+
+          if (type === types.BLOCK || type === types.MERKLE) {
+            position = reader.offset;
+
+            const headers = Headers.fromReader(reader);
+            hash = headers.hash();
+
+            reader.seek(length - consensus.HEADER_SIZE);
+          } else {
+            hash = reader.readHash();
+            position = reader.offset;
+            reader.seek(length);
+          }
+        } catch (err) {
+          this.logger.warning(
+            'Unknown block in file: %s, reason: %s',
+            filepath, err.message);
+          continue;
+        }
+
+        const blockrecord = new BlockRecord({
+          file: fileno,
+          position: position,
+          length: length
+        });
+
+        blocks += 1;
+        b.put(layout.b.encode(type, hash), blockrecord.encode());
+      }
+
+      const filerecord = new FileRecord({
+        blocks: blocks,
+        used: reader.offset,
+        length: this.maxFileLength
+      });
+
+      b.put(layout.f.encode(type, fileno), filerecord.encode());
+
+      await b.write();
+
+      this.logger.info('Indexed %d blocks (file=%s).', blocks, filepath);
+    }
+  }
+
+  /**
+   * Compares the number of files in the directory
+   * with the recorded number of files. If there are any
+   * inconsistencies it will reindex all blocks.
+   * @private
+   * @returns {Promise}
+   */
+
+  async index() {
+    await this._index(types.BLOCK);
+    await this._index(types.UNDO);
+    await this._index(types.MERKLE);
+  }
+
+  /**
+   * This method ensures that both the block storage directory
+   * and index directory exist.
+   * before opening.
+   * @returns {Promise}
+   */
+
+  async ensure() {
+    return fs.mkdirp(this.indexLocation);
+  }
+
+  /**
+   * Opens the file block store. It will regenerate necessary block
+   * indexing if the index is missing or inconsistent.
+   * @returns {Promise}
+   */
+
+  async open() {
+    this.logger.info('Opening FileBlockStore...');
```

Comment:
nitpick: this log and it's corresponding closing log looks a bit inconsistent. Can it say 'Opening/Closing block store' instead?

Reply:
I don't get it: both say `Opening FileBlockStore...` and `Closing FileBlockStore...`, which I believe was used for consistency with WalletDB and ChainDB logging.

Reply:
I guess that's OK. Was concerned it's a little bit too technical to go into logs as `info`.